### PR TITLE
Add new JSON reporter

### LINKED ISSUE #3504: Exporting to JSON does not honor score option
### Steps to reproduce
1. Run pylint on some random Python file or module:
```
pylint  ~/Desktop/pylint_test.py
```
As you can see this outputs some warnings/scoring:
```
************* Module pylint_test
/home/administrator/Desktop/pylint_test.py:1:0: C0111: Missing module docstring (missing-docstring)
/home/administrator/Desktop/pylint_test.py:1:0: W0611: Unused import requests (unused-import)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```
2. Now run the same command but with `-f json` to export it to JSON:
```
pylint ~/Desktop/pylint_test.py  -f json
```
The output doesn't contain the scores now anymore:
```
[
    {
        "type": "convention",
        "module": "pylint_test",
        "obj": "",
        "line": 1,
        "column": 0,
        "path": "/home/administrator/Desktop/pylint_test.py",
        "symbol": "missing-docstring",
        "message": "Missing module docstring",
        "message-id": "C0111"
    },
    {
        "type": "warning",
        "module": "pylint_test",
        "obj": "",
        "line": 1,
        "column": 0,
        "path": "/home/administrator/Desktop/pylint_test.py",
        "symbol": "unused-import",
        "message": "Unused import requests",
        "message-id": "W0611"
    }
]
```

3. Now execute it with `-f json` again but also supply the `--score=y` option:
```
[
    {
        "type": "convention",
        "module": "pylint_test",
        "obj": "",
        "line": 1,
        "column": 0,
        "path": "/home/administrator/Desktop/pylint_test.py",
        "symbol": "missing-docstring",
        "message": "Missing module docstring",
        "message-id": "C0111"
    },
    {
        "type": "warning",
        "module": "pylint_test",
        "obj": "",
        "line": 1,
        "column": 0,
        "path": "/home/administrator/Desktop/pylint_test.py",
        "symbol": "unused-import",
        "message": "Unused import requests",
        "message-id": "W0611"
    }
]
```

### Current behavior
The score is not outputted when exporting to JSON, not even when `--score=y` is activated.

### Expected behavior
The score is added to the JSON, at least when `--score=y` is activated.

### pylint --version output
```
pylint 2.3.0
astroid 2.2.0
Python 3.7.5 (default, Nov 20 2019, 09:21:52) 
[GCC 9.2.1 20191008]
```

### LINKED ISSUE #4741: Improve JSON output format with score
### Is your feature request related to a problem? Please describe

I want to consume this programmatically, but the format make it difficult since the JSON array will contain different types of objects:

```json
[
    {
        "type": "convention",
        "module": "fib",
        "obj": "",
        "line": 1,
        "column": 0,
        "path": "fib.py",
        "symbol": "missing-module-docstring",
        "message": "Missing module docstring",
        "message-id": "C0114"
    },
    {
        "score": "Your code has been rated at 3.33/10"
    }
]
```

### Describe the solution you'd like

Something like this would be better:

```json
{
    "score": 0.333,
    "violations": [
        {
            "type": "convention",
            "module": "fib",
            "obj": "",
            "line": 1,
            "column": 0,
            "path": "fib.py",
            "symbol": "missing-module-docstring",
            "message": "Missing module docstring",
            "message-id": "C0114"
        }
    ]
}
```

### Additional context

The score in the JSON output was added in https://github.com/PyCQA/pylint/pull/3514 for version `3.0.0a0`.

### PR DESCRIPTION
**Pull request by DanielNoord**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description

Closes #4741
Closes #3504

As discussed in the issue we create a new `json2` reporter that users can now use to get the score and other statistics. Since I want to keep being conscious about what is and isn't public API I decided to put both reporters in the private namespace. Since we are still in beta that should be fine? 😄

### CODE REVIEW

**Comment by DanielNoord** [comment]
@jacobtylerwalls Is the issue on `PyPy` related to changes to how we invoke tests? The module is clearly there 😅

File: doc/whatsnew/fragments/4741.feature [hunk 1]
```diff
@@ -0,0 +1,10 @@
+A new `json2` reporter has been added. It features a more enriched output that is
+easier to parse and provides more info.
+
+Compared to `json` the only changes are that messages are now under the `"messages"`
+key and that `"message-id"` now follows the camelCase convention and is renamed to
+`"messageId"`.
+
```
Comment (Pierre-Sassoulas) [inline doc/whatsnew/fragments/4741.feature:7]: There's something about the score to announce too, this is what prompted the change in the first place.

File: pylint/lint/base_options.py [hunk 2]
```diff
@@ -103,7 +103,7 @@ def _make_linter_options(linter: PyLinter) -> Options:
                 "short": "f",
                 "group": "Reports",
                 "help": "Set the output format. Available formats are text,"
-                " parseable, colorized, json and msvs (visual studio)."
+                " parseable, colorized, json2 and msvs (visual studio)."
```
Comment (Pierre-Sassoulas) [inline pylint/lint/base_options.py:106]: I don't think we should hide the fact that the json reporter still exists, generating this text would be nice too imo.

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Some preliminary comments, I'll do a more thorough review later (I'm on mobilel) I had only a quick look but nothing seems to be wrong

File: pylint/lint/base_options.py (hunk 2 above)
Comment (DanielNoord) [reply pylint/lint/base_options.py:106]: I guess we have a different opinion here then.

I think there should be one way to do things and that `json2` is superior. Thus, that option is what we should advertise to users looking through the possibilities.

**Comment by jacobtylerwalls** [comment]
> @jacobtylerwalls Is the issue on PyPy related to changes to how we invoke tests?

I don't know, I'd appreciate help investigating.

>  The module is clearly there 😅

From recent experience, no, nothing about package layouts without `src`-layout is clear ...! 😆

File: doc/whatsnew/fragments/4741.feature [hunk 3]
```diff
@@ -0,0 +1,10 @@
+A new `json2` reporter has been added. It features a more enriched output that is
```
Comment (jacobtylerwalls) [inline doc/whatsnew/fragments/4741.feature:1]: Double-backticks for code literals unless you truly wanted italics.

File: doc/whatsnew/fragments/4741.feature [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+easier to parse and provides more info.
+
+Compared to `json` the only changes are that messages are now under the `"messages"`
```
Comment (jacobtylerwalls) [inline doc/whatsnew/fragments/4741.feature:4]: same

File: pylint/lint/base_options.py (hunk 2 above)
Comment (jacobtylerwalls) [reply pylint/lint/base_options.py:106]: Yes, but imagine you maintain a tool that consumes the old reporter. And you're reviewing a PR, and you need to brush up on the context. You come here, and it's gone. Won't you burn time trying to figure out what happened and wondering why things aren't horribly broken?

File: pylint/reporters/_json_reporter.py [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+    name = "json"
+    extension = "json"
+
+    def display_messages(self, layout: Section | None) -> None:
+        """Launch layouts display."""
+        warnings.warn(
+            "The 'json' reporter has been deprecated in favour of a new version. "
+            "Consider switching to 'json2' which provides the same info in an updated "
+            "and improved format.",
+            DeprecationWarning,
```
Comment (jacobtylerwalls) [inline pylint/reporters/_json_reporter.py:56]: I don't think we have consensus to remove this just yet.
```suggestion
            PendingDeprecationWarning,
```

File: pylint/testutils/_primer/primer_command.py [hunk 6]
```diff
@@ -9,7 +9,7 @@
 from pathlib import Path
 from typing import Dict, TypedDict
 
-from pylint.reporters.json_reporter import OldJsonExport
+from pylint.reporters._json_reporter import OldJsonExport
```
Comment (jacobtylerwalls) [inline pylint/testutils/_primer/primer_command.py:12]: Good, I'm glad we're not immediately updating the primer to use json2. I have a toy branch that has a ShortJSONReporter that only captures the absolute minimum for the compare command to reduce file size.

File: tests/reporters/unittest_reporting.py [hunk 7]
```diff
@@ -208,37 +208,7 @@ def test_multi_format_output(tmp_path: Path) -> None:
...
-            '        "endColumn": null,\n'
-            f'        "path": {escaped_source_file},\n'
-            '        "symbol": "line-too-long",\n'
-            '        "message": "Line too long (1/2)",\n'
-            '        "message-id": "C0301"\n'
-            "    }\n"
-            "]\n"
-            "direct output\n"
-        )
+        assert '"messageId": "C0114"' in f.read()
```
Comment (jacobtylerwalls) [inline tests/reporters/unittest_reporting.py:211]: Why shorten the result?

**Review by jacobtylerwalls — COMMENTED** [review]
Big thanks for picking this up!

File: pylint/lint/base_options.py (hunk 2 above)
Comment (DanielNoord) [reply pylint/lint/base_options.py:106]: Isn't that what the changelog is for? If you search for `json` you'll find the change.

_I'm fine with adding something like `json (deprecated)` but as expressed elsewhere I would really like to nudge all users to the new reporter._

File: pylint/reporters/_json_reporter.py (hunk 5 above)
Comment (DanielNoord) [reply pylint/reporters/_json_reporter.py:56]: 👍  But let's keep the discussion going in the issue.

File: pylint/testutils/_primer/primer_command.py (hunk 6 above)
Comment (DanielNoord) [reply pylint/testutils/_primer/primer_command.py:12]: 👍 

Shall we open an issue to do so? It should be trivial to change your `ShortJSONReporter` to use `JSONReporterTwo` 😄

File: pylint/lint/base_options.py (hunk 2 above)
Comment (jacobtylerwalls) [reply pylint/lint/base_options.py:106]: Yeah I'm waffling on this one. As long as it's somewhere I guess it doesn't have to be in the CLI help.

File: pylint/testutils/_primer/primer_command.py (hunk 6 above)
Comment (jacobtylerwalls) [reply pylint/testutils/_primer/primer_command.py:12]: Oh, but I would want to keep the separate/private tiny one. (This one's too bulky).

Comment (DanielNoord) [reply pylint/testutils/_primer/primer_command.py:12]: But wouldn't you just need to use `"messages"` and then be done?

File: tests/reporters/unittest_reporting.py (hunk 7 above)
Comment (DanielNoord) [reply tests/reporters/unittest_reporting.py:211]: This doesn't need to test the full output, we only need to test that the output contains `"json"` output. That's the intent of the test.

If we want to test the format we should do so in the other tests that are specific to the json reporter (which this one is not).

File: pylint/testutils/_primer/primer_command.py (hunk 6 above)
Comment (jacobtylerwalls) [reply pylint/testutils/_primer/primer_command.py:12]: I was slicing away parts of the message, just keeping the necessary bits. Sometimes the message text is quite bulky.

File: tests/reporters/unittest_reporting.py (hunk 7 above)
Comment (jacobtylerwalls) [reply tests/reporters/unittest_reporting.py:211]: Thanks for explaining!

**Comment by jacobtylerwalls** [comment]
> Since I want to keep being conscious about what is and isn't public API I decided to put both reporters in the private namespace. Since we are still in beta that should be fine? 😄

My initial inclination is that we need a stronger reason to do this, but I'm not certain yet. Is there a reason we need to be this paranoid? Don't we anticipate never touching this again?

**Comment by DanielNoord** [comment]
> > Since I want to keep being conscious about what is and isn't public API I decided to put both reporters in the private namespace. Since we are still in beta that should be fine? 😄
> 
> My initial inclination is that we need a stronger reason to do this, but I'm not certain yet. Is there a reason we need to be this paranoid? Don't we anticipate never touching this again?

After seeing the trouble `CPython` is having with their C API and our own troubles with deprecating stuff I'm (currently) inclined to always make things private unless explicitly asked for. There is a good reason `pytest` has almost all of their package in a private namespace:
https://github.com/pytest-dev/pytest/tree/main/src/pytest

Whenever somebody want to access something I'm highly in favour of making it public and creating a well-defined API for it, but I just don't want to deal with another spree of deprecation and removal issues.

**Comment by cdce8p** [comment]
> > My initial inclination is that we need a stronger reason to do this, but I'm not certain yet. Is there a reason we need to be this paranoid? Don't we anticipate never touching this again?
> 
> After seeing the trouble `CPython` is having with their C API and our own troubles with deprecating stuff I'm (currently) inclined to always make things private unless explicitly asked for. There is a good reason `pytest` has almost all of their package in a private namespace: https://github.com/pytest-dev/pytest/tree/main/src/pytest
> 
> Whenever somebody want to access something I'm highly in favour of making it public and creating a well-defined API for it, but I just don't want to deal with another spree of deprecation and removal issues.

It it really necessary to add an underscore? Yes deprecations are quite a hassle, but it also has it's upsides. E.g. if users want to customize any of the existing reporters, they can easily subclass it. As long as it isn't documented publicly, it's just kindness / good responsibility of us not to break stuff unnecessarily.

Another example, consider the `checkers/utils.py` module which is probably used by a lot of plugins. We can still change things. Plugins authors will adjust. Only it shouldn't be for `x.1`, then `x.2`, `x.3` and so on. Change it once and stick with it (at least for some time).

**Comment by DanielNoord** [comment]
> E.g. if users want to customize any of the existing reporters, they can easily subclass it. As long as it isn't documented publicly, it's just kindness / good responsibility of us not to break stuff unnecessarily.

If there is a need for it they can for it to become public and we can then make that explicit. For me using the underscores is also about setting expectations about what can and what can not change. I'm fine with maintaining a large public API, but only when we see a need for it because plugin writers or users ask for it. From the current state we are in we can only get there by making new code explicitly private.

> Another example, consider the `checkers/utils.py` module which is probably used by a lot of plugins. We can still change things. Plugins authors will adjust. Only it shouldn't be for `x.1`, then `x.2`, `x.3` and so on. Change it once and stick with it (at least for some time).

Again, I think there is too much of a grey array there. There were quite a bit of number of functions that I would have liked to change or delete when doing the configuration rewrite and the typing stuff but which I couldn't. Some `utils` should just be for internal use and should be documented as such.

File: doc/whatsnew/fragments/4741.feature [hunk 8]
```diff
@@ -0,0 +1,12 @@
+A new ``json2`` reporter has been added. It features a more enriched output that is
+easier to parse and provides more info.
+
+Compared to ``json`` the only changes are that messages are now under the ``"messages"``
+key and that ``"message-id"`` now follows the camelCase convention and is renamed to
+``"messageId"``.
+The new reporter also reports the "score" of the modules you linted as defined by the
+``evaluation`` option and provides statistics about the modules you linted.
```
Comment (Pierre-Sassoulas) [inline doc/whatsnew/fragments/4741.feature:8]: I think we should just dump a json example. Descriptions of data structures are hard to follow at least for me. The minitmal example is this right ?
```json
{
    "messages": [
        {
            "type": "convention",
            "symbol": "line-too-long",
            "message": "Line too long (1/2)",
            "messageId": "C0301",
            "confidence": "HIGH",
            "module": "0123",
            "obj": "",
            "line": 1,
            "column": 0,
            "endLine": 1,
            "endColumn": 4,
            "path": "0123",
            "absolutePath": "0123"
        }
    ],
    "statistics": {
        "messageTypeCount": {
            "fatal": 0,
            "error": 0,
            "warning": 0,
            "refactor": 0,
            "convention": 1,
            "info": 0
        },
        "modulesLinted": 1,
        "score": 5.0
    }
}

```

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]

**Review by jacobtylerwalls — CHANGES_REQUESTED** [review]
I think I agree with Marc that we should prefer building new reporters like json3 if it ever came to that rather than make this private.

File: doc/whatsnew/fragments/4741.feature [hunk 9]
```diff
@@ -0,0 +1,12 @@
+A new ``json2`` reporter has been added. It features a more enriched output that is
+easier to parse and provides more info.
```
[the remark this answers is not in the document]
Comment (DanielNoord) [reply doc/whatsnew/fragments/4741.feature:2]: This is already included in a later paragraph

File: doc/whatsnew/fragments/4741.feature (hunk 8 above)
Comment (DanielNoord) [reply doc/whatsnew/fragments/4741.feature:8]: I can't get this to pass the `check-news fragment` check. I'd say let's skip it for now, the time investment necessary to fix it seems to be quite high for little gain.

**Comment by DanielNoord** [comment]
I rebased and force pushed to remove the private making commits.

This now only introduce a new reporter and uses it where necessary/possible. Let me know what you guys think!

File: doc/whatsnew/fragments/4741.feature (hunk 8 above)
Comment (jacobtylerwalls) [reply doc/whatsnew/fragments/4741.feature:8]: Fair. You could also edit the 3.0.md directly and skip making a fragment.

Comment (DanielNoord) [reply doc/whatsnew/fragments/4741.feature:8]: Smart! Let's do that!

File: tests/reporters/unittest_json_reporter.py [hunk 10]
```diff
@@ -132,6 +133,133 @@ def get_linter_result(score: bool, message: dict[str, Any]) -> list[dict[str, An
...
+            id="not-everything-defined",
+        ),
+    ],
+)
+def test_serialize_deserialize_for_v2(message: Message) -> None:
+    json_message = JSON2Reporter.serialize(message)
+    assert message == JSON2Reporter.deserialize(json_message)
+
+
+def test_json2_result_wiuth_broken_score() -> None:
```
Comment (jacobtylerwalls) [inline tests/reporters/unittest_json_reporter.py:257]:
```suggestion
def test_json2_result_with_broken_score() -> None:
```

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Look ready, it only needs the example of json in 3.0.md that Jacob suggested and let's merge 👍

File: doc/whatsnew/fragments/4741.feature (hunk 8 above)
Comment (DanielNoord) [reply doc/whatsnew/fragments/4741.feature:8]: Can we do that already @Pierre-Sassoulas or should we wait until the final release notes are written?

Comment (Pierre-Sassoulas) [reply doc/whatsnew/fragments/4741.feature:8]: I think we can start writing a draft of the release notes, it's going to be easier if we do it progressively.

**Review by jacobtylerwalls — APPROVED** [review]

**Review by Pierre-Sassoulas — APPROVED** [review]
